app/models/concerns/smart_proxy_reference.rb

```diff
@@ -0,0 +1,51 @@
+module SmartProxyReference
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    attr_accessor :smart_proxy_references
+
+    def smart_proxy_reference(hash)
+      raise Foreman::Exception.new(N_("Proxy reference should be { :relation => [:column_name, ...]}")) unless correct_reference_format?(hash)
+      @smart_proxy_references ||= {}
+      @smart_proxy_references.merge!(hash)
+    end
+
+    def proxy_self_columns
+      proxy_join_references(true).values.first.map(&:to_s)
+    end
+
+    def proxy_join_tables
+      proxy_join_references.keys
+    end
+
+    def create_groups
+      proxy_join_references.inject([]) { |memo, (k, v)| memo << { k => v } }
+        .group_by { |h| Host::Managed.reflections[h.keys.first.to_s].table_name }.with_indifferent_access
+    end
+
+    def proxy_column_list
+      columns = create_groups.map do |table_name, relations|
+        relations.flat_map.with_index do |relation, i|
+          relation.map do |rel_name, columns|
+            columns.map do |col_name|
+              rename = i > 0 ? "_hosts" : ""
+              "#{rel_name.pluralize}#{rename}.#{col_name}"
```

Comment:
In case of the first relation, table name would be taken, not the relation one:

``` ruby
alias_name = i > 0 ? "#{rel_name.pluralize}_hosts" : table_name
"#{alias_name}.#{col_name}"
```